src/world/World.php

```diff
@@ -1977,14 +2029,21 @@ public function getBlockAt(int $x, int $y, int $z, bool $cached = true, bool $ad
 			$chunk = $this->chunks[$chunkHash] ?? null;
 			if($chunk !== null){
 				$block = $this->blockStateRegistry->fromStateId($chunk->getBlockStateId($x & Chunk::COORD_MASK, $y, $z & Chunk::COORD_MASK));
+				$displacedBlockStateId = $chunk->getDisplacedBlockStateId($x & Chunk::COORD_MASK, $y, $z & Chunk::COORD_MASK);
 			}else{
 				$addToCache = false;
 				$block = VanillaBlocks::AIR();
+				$displacedBlockStateId = Block::EMPTY_STATE_ID;
 			}
 		}else{
 			$block = VanillaBlocks::AIR();
+			$displacedBlockStateId = Block::EMPTY_STATE_ID;
 		}
 
+		if($block instanceof Waterloggable && $displacedBlockStateId !== Block::EMPTY_STATE_ID){
+			$displacedBlockStateId = $this->blockStateRegistry->fromStateId($displacedBlockStateId);
+			$block->setContainedWater($displacedBlockStateId instanceof Water ? $displacedBlockStateId : null);
+		}
```

Comment:
I'm thinking we could probably use `readStateFromWorld()` to do this, instead of baking the logic into `getBlock` directly